### ISSUE TITLE
Circumventing Linearity

### ISSUE THREAD

**Issue by buggymcbugfix** [granule-project/granule#54]
There is something funny going on with `filter` below: we are applying `predicate` to `x` and then using `x` in one branch but not the other. By virtue of linearity we shouldn't allow it to be used in either branch.

```idris
import Prelude
import Sequence

filter : (Int -> Bool) |∞| -> Sequence Int -> Sequence Int
filter |predicate| list =
    case list of
      Empty -> Empty;
      (Next x xs) -> if predicate x
        then Next x (filter |predicate| xs)
        else filter |predicate| xs

main : Sequence Int
main = filter |(\(x : Int) -> x < 3)| (Next 1 (Next 2 (Next 3 (Next 4 Empty))))
```

**Comment by dorchard** [comment]
This is probably a bug in synthesising the type of `case` - I am looking into it.

**Comment by dorchard** [comment]
Fixed by 9654ccb854eac9b5f8b711ebc844acb8ecb72e09 (and added as a negative regression)
